src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -774,54 +771,41 @@ vector<SparqlTripleSimpleWithGraph> Visitor::visit(
 }
 
 // ____________________________________________________________________________________
-vector<SparqlTripleSimpleWithGraph> Visitor::transformTriplesTemplate(
-    Parser::TriplesTemplateContext* ctx,
-    const SparqlTripleSimpleWithGraph::Graph& graph) {
-  auto convertTriple = [&graph](const std::array<GraphTerm, 3>& triple)
-      -> SparqlTripleSimpleWithGraph {
-    return {visitGraphTerm(triple[0]), visitGraphTerm(triple[1]),
-            visitGraphTerm(triple[2]), graph};
-  };
-
-  return ad_utility::transform(visit(ctx), convertTriple);
-}
-
-// ____________________________________________________________________________________
-vector<SparqlTripleSimpleWithGraph> Visitor::visit(Parser::QuadsContext* ctx) {
+Quads Visitor::visit(Parser::QuadsContext* ctx) {
   // The ordering of the individual triplesTemplate and quadsNotTriples is not
   // relevant and also not known.
-  auto triplesWithGraph = ad_utility::transform(
-      ctx->triplesTemplate(), [this](Parser::TriplesTemplateContext* ctx) {
-        return transformTriplesTemplate(ctx, std::monostate{});
-      });
-  ql::ranges::move(visitVector(ctx->quadsNotTriples()),
-                   std::back_inserter(triplesWithGraph));
-  return ad_utility::flatten(std::move(triplesWithGraph));
+  Quads quads;
+  quads.freeTriples_ = ad_utility::flatten(visitVector(ctx->triplesTemplate()));
+  for (auto [graph, triples] : visitVector(ctx->quadsNotTriples())) {
+    quads.graphTriples_.emplace_back(std::move(graph), std::move(triples));
+  }
+  return quads;
 }
 
 // ____________________________________________________________________________________
-vector<SparqlTripleSimpleWithGraph> Visitor::visit(
-    Parser::QuadsNotTriplesContext* ctx) {
-  // Short circuit when the triples section is empty
-  if (!ctx->triplesTemplate()) {
-    return {};
-  }
-
+Quads::GraphBlock Visitor::visit(Parser::QuadsNotTriplesContext* ctx) {
   auto graphTerm = visit(ctx->varOrIri());
-  SparqlTripleSimpleWithGraph::Graph graph = graphTerm.visit(
-      [&ctx]<typename T>(
-          const T& element) -> SparqlTripleSimpleWithGraph::Graph {
-        if constexpr (std::is_same_v<T, Variable> || std::is_same_v<T, Iri>) {
+  Quads::IriOrVariable graph = graphTerm.visit(
+      [&ctx]<typename T>(const T& element) -> Quads::IriOrVariable {
+        if constexpr (std::is_same_v<T, Variable>) {
           return element;
+        } else if constexpr (std::is_same_v<T, Iri>) {
+          return TripleComponent::Iri::fromIriref(element.iri());
         } else {
           static_assert(std::is_same_v<T, BlankNode> ||
                         std::is_same_v<T, Literal>);
           reportError(ctx->varOrIri(),
-                      "Only IRIs and variables are allowed as graph names.");
+                      "Internal Error: got unexpected type. Please report this "
+                      "to the developers.");
```

Comment:
Is the reason here that `visit(varOrIri)` doesn't return `varOrIri` but something more general that can also store additional types?
Then write a comment here, otherwise this looks a bit confusing.

Reply:
I see no reason for this. This goes into a similar direction as unifying the two Iri types. Downstream of `VarOrIri` different Iri types are required which makes it somewhat awkard.

Reply:
I am sorry, but I don't quite understand your comment.
So the assesment "The used types here are a bit messy/too loose, so we unfortunately need this for now" which should be stated by a comment remains?